accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,504 @@
+
+For the first two, we could put the static members on the `Iterable` class, but since the extension does not apply to *all* iterables, it is not clear that this is correct.
+
+For `int Function(int)` and `FutureOr<int>`, it's unclear how to call such a static method at all. We can denote`int Function(int)` with a type alias, but putting static members on type aliases is a new concept.  We could put the static method on `Function`, but that's not particularly discoverable, and why not require that they are put on `Functon` explicitly. For `FutureOr`, we could allow static members on `FutureOr` (which is a denotable type), but again it seems spurious. For `int?`, we could put the method on `int`, but why not  just require that it's on `int`.
+
+The issue here is that the type patterns used by `on` are much more powerful than what is necessary to put static members on class types.
+
+It would probably be more readable to introduce a proper static extension declaration:
+
+```dart
+static extension Foo on int {  // or: extension Foo on static int 
+  int fromList(List l) => l.length;
+}
+
+...
+  print(int.fromList([1, 2])); // 2
+```
+
+where the `on` type must be something that can already have static methods. 
+
+The disadvantage is that if you want to introduce related functionality that is both static and instance methods on a class, then you need to write two extensions with different names.
+
+If we allow extension static declarations like these, we can also allow extension constructors.
+
+### Explicitly Specify Related Declarations
+
+The above resolution strategy uses an implicit "more specific" relation between applicable extensions to select one of them (when possible). This may cause surprises when two unrelated extensions happen to both apply.
+
+Alternatively, we can allow extensions to declare that they are *related*, and only allow conflicts between related extensions, which are presumably aware of each other.
+
+The syntax would be:
+
+```dart
+extension Foo<T> on Iterable<T> { twizzle() {...} fromp() { ... }}
+extension Bar<T> extends Foo<T> on List<T> { twizzle() { ... } } 
+```
+
+If both of these extensions apply, then pick the one that extends the other. If there are more related extensions which apply, then pick one which transitively extends all the others.
+
+If there isn't exactly one among applicable extensions which extends all the rest, the conflict is a compile-time error.
+
+This approach allows related extensions to declare functionality on a number of types, without accidentally allowing a conflict with an unrelated extension.
+
+The extending extension must be defined on a subtype of its super-extension. In the above, the `on` type of `Bar<T>` is `List<T>`, which is a subtype of the `on` type of `Foo<T>`, which is `Iterable<T>`. Also, the declared extension methods in the extending extension must be valid overrides of the same-named super-extensions extension methods, as if it was a subclass relationship.
```

Comment:
Do you mean subtype in general, or proper subtype?  Can I do
```
extension Baz<T> extends Foo<T> on Iterable<T> { twizzle() { ... }}
```

What about

```
extension A<T> on T { twizzle() { ... }}
extension B<T> extends A<T> on T { twizzle() { ... } }
```

The latter would give you a way to override completely generic extensions, which we wouldn't otherwise have, I think?

Reply:
Just subtype. You *can* completely shadow the super-extension, but I don't think avoiding that isn't worth much effort.